llama : refactor model loading code

ggerganov opened the pull request:
- extend `gguf` API for writing
- simplify `llama.cpp` model loading / saving by using new `gguf` API
- reorder definitions in `llama.cpp` to match interface order
- merge `utils.h` in `llama.cpp`

ggerganov committed:
llama : style formatting + remove helper methods

ggerganov committed:
llama : fix quantization using gguf tool

ggerganov commented:
@monatis  The quantization works now - I had messed up the function overloads for writing arrays / strings

ggerganov committed:
llama : simplify gguf_file_saver

ggerganov committed:
llama : fix method names

ggerganov committed:
llama : simplify write_header()

monatis commented:
>  The quantization works now 

I think we also need to implement writing of int arrays after adding token types yesterday

ggerganov committed:
llama : no need to pass full file loader to the file saver

just gguf_ctx

ggerganov committed:
llama : gguf_file_saver write I32

monatis committed:
llama : refactor tensor names (#2622)

* gguf: update tensor names searched in quantization

* gguf : define tensor names as constants

ggerganov committed:
gguf : initial write API (not tested yet)

ggerganov committed:
gguf : write to file API (not tested)

ggerganov committed:
gguf : initial write API ready + example

ggerganov committed:
gguf : fix header write

ggerganov committed:
gguf : fixes + simplify example + add ggml_nbytes_pad()

ggerganov committed:
gguf : minor

ggerganov committed:
llama : replace gguf_file_saver with new gguf write API

slaren commented:
I am not sure if this is the right place to do this, but I think that at some point we should split `llama_context_params` into `llama_model_params` and `llama_context_params`, to make clear what parameters are used when loading a model, and what parameters can be modified per-context. For example, `use_mmap` is a model loading parameter and should be part of `llama_model_params`, but `n_ctx` should be configurable per-context in `llama_context_params`. As it is now, the users must assume that `llama_load_model_from_file` and `llama_new_context_with_model` must receive the exact same parameters in `llama_context_params`, as it is not clear which ones are relevant for each function.

ggerganov committed:
gguf : streaming support when writing files

ggerganov committed:
gguf : remove oboslete write methods

ggerganov committed:
gguf : remove obosolete gguf_get_arr_xxx API

ggerganov committed:
llama : simplify gguf_file_loader

ggerganov committed:
llama : move hparams and vocab from gguf_file_loader to llama_model_loader

ggerganov committed:
llama : merge gguf-util.h in llama.cpp

ggerganov committed:
llama : reorder definitions in .cpp to match .h

ggerganov committed:
llama : minor simplifications

klosax commented on gguf-llama.cpp:
```diff
@@ -1256,15 +1292,33 @@ static void llama_model_load_internal(
+        hparams.n_head         = gguf_get_val_u32(ctx, gguf_find_key(ctx, "llama.attention.head_count"));
+        hparams.n_layer        = gguf_get_val_u32(ctx, gguf_find_key(ctx, "llama.block_count"));
+        hparams.n_rot          = gguf_get_val_u32(ctx, gguf_find_key(ctx, "llama.rope.dimension_count"));
+        hparams.f_rms_norm_eps = gguf_get_val_f32(ctx, gguf_find_key(ctx, "llama.rms_norm_epsilon"));
```
the key for rms norm is `llama.attention.layer_norm_rms_epsilon` in the spec and conversion scripts

ggerganov replied:
Ah, will fix it here - thanks for pointing out

ggerganov committed:
llama : refactor llama_model_loader (WIP)

wip : remove ggml_ctx from llama_model_loader

wip : merge gguf_file_loader in llama_model_loader

monatis commented:
I thought that we could reuse some of the code parts  for other LLMs, moving them to a common lib, but now I have the impression that things are going to a kinda self-contained structure. Is this a temporary state, or is the plan to isolate implementations of different LLMs?

ggerganov commented:
My plan currently is to have `llama.cpp` implement the things that are common for all models. Ideally, when we add a new model we'll just need to implement a corresponding `struct llama_model` and `llama_build_graph()` function - these potentially can be separated in different source files, but I need to first have an example running to make a decision how to reorganize best.

The `gguf-utils.h` (formerly `llama-utils.h`) has always been "out of place". I tried merging it in `ggml`, but it does not fit really well. Maybe in the future. For now, I don't see a reason to not be inside `llama.cpp`. The `mmap` and `mlock` functionality will be shared by all models.

ggerganov committed:
llama : fix shape prints

ggerganov commented:
Haven't tested GPU and LoRA parts 

Plan is to merge this and then de-duplicate the `gguf-llama.cpp` and `llama.cpp` and make sure that all features work

monatis commented:
clang++ on WSL fails to build without including `cstdarg` (for `format()`).

ggerganov commented:
> clang++ on WSL fails to build without including `cstdarg` (for `format()`).

Thanks - will fix it now.

Can you confirm that the following works as expected:

```bash
# run F16 inference
./gguf-llama-simple models/7B-v2/ggml-model-f16.gguf

# quantize
./gguf ./models/7B-v2/ggml-model-f16.gguf q
mv quant.gguf models/7B-v2/ggml-model-q4_0.gguf

# run quantized inference
./gguf-llama-simple models/7B-v2/ggml-model-q4_0.ggu
```

(replace with your respective paths

ggerganov committed:
llama : fix Windows build + fix norm_rms_eps key

ggerganov committed:
llama : throw error on missing KV paris in model meta data

ggerganov committed:
llama : improve printing + log meta data

ggerganov committed:
llama : switch print order of meta data
